### ISSUE TITLE
(regression) ICE compiling string_cache v0.2.29

### ISSUE THREAD

**Issue by crumblingstatue** [rust-lang/rust#40469]
```
Compiling string_cache v0.2.29
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/option.rs:323
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: rustc_metadata::cstore::CStore::get_crate_data
   1: rustc_metadata::cstore_impl::<impl rustc::middle::cstore::CrateStore for rustc_metadata::cstore::CStore>::def_key
   2: rustc_resolve::build_reduced_graph::<impl rustc_resolve::Resolver<'a>>::macro_def_scope
   3: rustc_resolve::Resolver::resolve_path
   4: rustc_resolve::Resolver::resolve_qpath
   5: rustc_resolve::Resolver::smart_resolve_path_fragment
   6: rustc_resolve::Resolver::smart_resolve_path
   7: rustc_resolve::Resolver::resolve_expr
   8: rustc_resolve::Resolver::resolve_block
   9: <rustc_resolve::Resolver<'a> as syntax::visit::Visitor<'tcx>>::visit_fn
  10: syntax::visit::walk_impl_item
  11: rustc_resolve::Resolver::with_type_parameter_rib
  12: rustc_resolve::Resolver::with_current_self_type
  13: rustc_resolve::Resolver::with_self_rib
  14: rustc_resolve::Resolver::with_optional_trait_ref
  15: rustc_resolve::Resolver::with_self_rib
  16: rustc_resolve::Resolver::resolve_item
  17: syntax::visit::walk_item
  18: rustc_resolve::Resolver::resolve_item
  19: rustc_resolve::Resolver::resolve_crate
  20: rustc_driver::driver::phase_2_configure_and_expand
  21: rustc_driver::driver::compile_input
  22: rustc_driver::run_compiler
  23: std::panicking::try::do_call
  24: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:98
  25: <F as alloc::boxed::FnBox<A>>::call_box
  26: std::sys::imp::thread::Thread::new::thread_start
             at /checkout/src/liballoc/boxed.rs:650
             at /checkout/src/libstd/sys_common/thread.rs:21
             at /checkout/src/libstd/sys/unix/thread.rs:84
  27: start_thread
  28: clone

error: Could not compile `string_cache`.

```
`rustc 1.17.0-nightly (824c9ebbd 2017-03-12)`

**Comment by SimonSapin** [comment]
Same on string-cache master (version 0.4.0+).

**Comment by jseyfried** [comment]
I believe this was caused by #40220 -- I'll fix ASAP.

**Comment by jseyfried** [comment]
Fixed in #40583.
